include/envoy/ssl/private_key/private_key.h

```diff
@@ -0,0 +1,83 @@
+#pragma once
+
+#include <functional>
+#include <string>
+
+#include "envoy/api/v2/auth/cert.pb.h"
+#include "envoy/common/pure.h"
+#include "envoy/event/dispatcher.h"
+#include "envoy/ssl/private_key/private_key_callbacks.h"
+
+#include "openssl/ssl.h"
+
+namespace Envoy {
+namespace Server {
+namespace Configuration {
+// Prevent a dependency loop with the forward declaration.
+class TransportSocketFactoryContext;
+} // namespace Configuration
+} // namespace Server
+
+namespace Ssl {
+
+typedef std::shared_ptr<SSL_PRIVATE_KEY_METHOD> BoringSslPrivateKeyMethodSharedPtr;
+
+class PrivateKeyConnection {
+public:
+  virtual ~PrivateKeyConnection() {}
+};
+
+typedef std::unique_ptr<PrivateKeyConnection> PrivateKeyConnectionPtr;
+
+class PrivateKeyMethodProvider {
+public:
+  virtual ~PrivateKeyMethodProvider() {}
+
+  /**
+   * Get a private key operations instance from the provider.
+   * @param ssl a SSL connection object.
+   * @param cb a callbacks object, whose "complete" method will be invoked
+   * when the asynchronous processing is complete.
+   * @param dispatcher supplies the owning thread's dispatcher.
+   * @return the private key operations instance.
+   */
+  virtual PrivateKeyConnectionPtr getPrivateKeyConnection(SSL* ssl,
```

Comment:
Why is this called `PrivateKeyConnection`? I.e. what is the connection aspect? Connection is a heavily overloaded term in Envoy. Also, what are the object lifetime issues here and the purpose of this method? Who uses it? Etc. Having some clarify around this would help improve the readability.

Reply:
The idea was that this part corresponds to a single connection. I agree the naming isn't very good here. I'll try to come up with a better name.

Reply:
The functionality could be contained in the provider. I remember thinking about that approach too, but this seemed bit more straightforward because of the memory management (it's not possible that the callback comes after the socket was destroyed) and easier thinking about thread data access in the provider. If we remove the `PrivateKeyConnection` object we would have an API something like this:
```c++
// throws an error if the registration fails
void registerPrivateKeyConnection(SSL* ssl, PrivateKeyConnectionCallbacks& cb, Event::Dispatcher& dispatcher) PURE;

void unregisterPrivateKeyConnection(SSL* ssl) PURE;
```
Internally the provider would need to keep track of registered SSL objects (which it just borrows from the SSL sockets). A benefit of this approach would be avoiding of the naming issue. :-) I think I'll experiment a bit with this to get a better idea how it will work.

Reply:
> it's not possible that the callback comes after the socket was destroyed

I don't think that's true (but I agree that the current model would be acceptable if it was).

In a security-oriented use case, the private key will be loaded in a separate process, or even on a separate host, which means that you can receive result of the computation long after the socket was destroyed.